src/webgpu/util/fp_interval.ts

```diff
@@ -0,0 +1,238 @@
+    assert(intervals.length > 0, `span of an empty list of FPIntervals is not allowed`);
+    const mins = new Set<number>();
+    const maxs = new Set<number>();
+    intervals.forEach(i => {
+      mins.add(i.begin);
+      maxs.add(i.end);
+    });
+    return new FPInterval(Math.min(...mins), Math.max(...maxs));
+  }
+
+  public toString(): string {
+    return `[${this.begin}, ${this.end}]`;
+  }
+}
+
+export interface FPUnaryIntervalBuilder {
+  singular(x: number): FPInterval;
+  range?(i: FPInterval): FPInterval;
+}
+
+export interface FPBinaryIntervalBuilder {
+  singular(x: number, y: number): FPInterval;
+  range?(ix: FPInterval, iy: FPInterval): FPInterval;
+}
+
+class AbsoluteFPIntervalBuilder implements FPUnaryIntervalBuilder {
+  protected readonly n: number;
+
+  public constructor(n: number) {
+    this.n = Math.abs(n);
+  }
+
+  public singular(x: number): FPInterval {
+    assert(!Number.isNaN(x), `absolute not defined for NaN`);
+    return FPInterval.span(this.impl(x), this.impl(flushSubnormalNumber(x)));
+  }
+
+  public range(i: FPInterval): FPInterval {
+    if (i.begin === i.end) {
+      return this.singular(i.begin);
+    }
+    return FPInterval.span(this.singular(i.begin), this.singular(i.end));
+  }
+
+  private impl(x: number): FPInterval {
+    return new FPInterval(x - this.n, x + this.n);
+  }
+}
+
+class CorrectlyRoundedFPIntervalBuilder implements FPUnaryIntervalBuilder {
+  public constructor() {}
+
+  public singular(x: number): FPInterval {
+    assert(!Number.isNaN(x), `correctlyRounded not defined for NaN`);
+    return FPInterval.span(this.impl(x), this.impl(flushSubnormalNumber(x)));
+  }
+
+  public range(i: FPInterval): FPInterval {
+    if (i.begin === i.end) {
+      return this.singular(i.begin);
+    }
+    return FPInterval.span(this.singular(i.begin), this.singular(i.end));
+  }
+
+  private impl(x: number): FPInterval {
+    if (x === Number.POSITIVE_INFINITY || x > kValue.f32.positive.max) {
+      return new FPInterval(kValue.f32.positive.max, Number.POSITIVE_INFINITY);
+    }
+
+    if (x === Number.NEGATIVE_INFINITY || x < kValue.f32.negative.min) {
+      return new FPInterval(Number.NEGATIVE_INFINITY, kValue.f32.negative.min);
+    }
+
+    const x_32 = new Float32Array([x])[0];
+    const converted: number = x_32;
+    if (x === converted) {
+      // x is precisely expressible as a f32, so correctly rounding degrades to exactly matching
+      return new FPInterval(x, x);
+    }
+
+    if (converted > x) {
+      // x_32 rounded towards +inf, so is after x
+      const otherside = nextAfter(x_32, false, false).value as number;
+      return new FPInterval(otherside, converted);
+    } else {
+      // x_32 rounded towards -inf, so is before x
+      const otherside = nextAfter(x_32, false, false).value as number;
+      return new FPInterval(converted, otherside);
+    }
+  }
+}
+
+class ULPFPIntervalBuilder implements FPUnaryIntervalBuilder {
```

Comment:
ULP will still need the test case generator to calculate n for certain cases based off of the inputs into the tests, but we already do that, so there is no increase in complexity.